### ISSUE TITLE
PubSub get_message() not always returns None when there're no data available

### ISSUE THREAD

**Issue by whiteplanet** [NoneGG/aredis#56]
Hello Jason,

Looks like PubSub get_message() function is not working as expected, on  just created channel without no data. I assume that call of this method should return None in that case, however None returns only once and then execution "hangs" and awaits for coroutine to be completed (actual data appears in the channel) and then execution continues.

Here are the steps to reproduce this issue:

```
# 1. Create some subscription (PubSub obj)

# redis — initialized connection to some Redis instance
pbsb = redis.pubsub(ignore_subscribe_messages=True)
await pbsb.subscribe("some-channel")

# 2. Query message in the loop
timeout = 5
message = None
end_time = time.time() + timeout
while timeout == 0 or time.time() < end_time:
    print(1)
    message = await pbsb.get_message(timeout=1)
    print(2)

# 3. results
>>> 1
>>> 2
>>> 1
# these should be a little more 1 and 2 :)

# 4. Nothing more happened until some data arrives, then
>>> 2
```

I've attached a [patch](https://github.com/NoneGG/aredis/files/1308292/get_message.patch.txt)
 addressed to resolve this issue, it removes `connection.can_read()` check in **aredis/pubsub.py** on line 132, because in that case it always returns `False` and adds several imports:
```
from aredis.exceptions import ConnectionError
from aredis.exceptions import TimeoutError
```
Looks like without them `except` on line 112 is not working as expected.

Also in **aredis/connection.py** on line 307 I've added check
```
if not self._reader:
    break
```

Because when coroutine actually cancelled, `self._reader` is getting None and this is leads to unwanted exception.

Best,
Jack

**Comment by NoneGG** [comment]
@whiteplanet 
Thank you for your bug report and patch(may be next time you can make a pull request directly~), that is a realization problem indeed.
But i think check on `can_read` is still necessary so i change the check to how redis-py do it. 
Your script can now be run well in my dev env, if you don't mind, could you test it by reinstall aredis from git?
`pip install git+https://github.com/NoneGG/aredis.git` 

Thank you again for the bug report, if there are still any other problems please let me know.

**Comment by whiteplanet** [comment]
Hi Jason,

Unfortunately it doesn't work for me. Right now I get None in return in case with empty channel. However when other side of my app pushes data into the channel, it doesn't appears on the side where get_message() actually called (get_message continue returning None).

Also, with this update I've noticed another issue when Debug is turned ON for the loop:
`Executing <Handle <TaskWakeupMethWrapper object at 0x7fa20aa0ac78> created at /usr/lib/python3.6/asyncio/futures.py:348> took 0.102 seconds`

Best,
Jack

**Comment by whiteplanet** [comment]
Hi Jason,

I've checked this code is always False for me.
```
await connection.can_read(timeout=timeout)
```

Also looks like this guy
```
bool(select([sock], [], [], timeout)[0])
```
Works really slow and raises warning I've previously mentioned.

Best,
Jack

**Comment by NoneGG** [comment]
Dear @whiteplanet 

## get_message always return None
I am not sure, `can_read` seems do have problem when there are msg published, i will check that and give a solution as soon as possible.

## about the TaskWakeupMethWrapper 
This seems to be the problem of `asyncio` rather than `aredis`, links below are also about that:
https://github.com/channelcat/sanic/issues/245
https://bugs.python.org/issue31096

About the speed, don't worry, it is just the time costs by what you put in `run_until_complete`, i run your script like below:
```Python
import asyncio
import logging
import time

import aredis


async def run():
    # 1. Create some subscription (PubSub obj)
    redis = aredis.StrictRedis()
    # redis — initialized connection to some Redis instance
    pbsb = redis.pubsub(ignore_subscribe_messages=True)
    await pbsb.subscribe("some-channel")

    # 2. Query message in the loop
    timeout = 5
    message = None
    start_at = time.time()
    end_time = start_at + timeout
    while timeout == 0 or time.time() < end_time:
        print(1)
        message = await pbsb.get_message(timeout=4)
        print(2)
    end_at = time.time()
    print(end_at - start_at)

if __name__ == '__main__':
    logging.basicConfig(level=logging.DEBUG)
    loop = asyncio.get_event_loop()
    loop.set_debug(enabled=True)
    loop.run_until_complete(run())
```
And the output is like:
```
/home/cm/work/aredis/env3.6/bin/python /home/cm/work/aredis/examples/script.py
DEBUG:asyncio:Using selector: EpollSelector
DEBUG:asyncio:Get address info localhost:6379, type=<SocketKind.SOCK_STREAM: 1>
1
DEBUG:asyncio:Getting address info localhost:6379, type=<SocketKind.SOCK_STREAM: 1> took 11.368 ms: [(<AddressFamily.AF_INET: 2>, <SocketKind.SOCK_STREAM: 1>, 6, '', ('127.0.0.1', 6379))]
DEBUG:asyncio:poll took 1.467 ms: 1 events
DEBUG:asyncio:connect <socket.socket fd=6, family=AddressFamily.AF_INET, type=2049, proto=6, laddr=('0.0.0.0', 0)> to ('127.0.0.1', 6379)
DEBUG:asyncio:poll took 0.005 ms: 1 events
DEBUG:asyncio:<socket.socket fd=6, family=AddressFamily.AF_INET, type=2049, proto=6, laddr=('127.0.0.1', 33482), raddr=('127.0.0.1', 6379)> connected to localhost:6379: (<_SelectorSocketTransport fd=6 read=polling write=<idle, bufsize=0>>, <asyncio.streams.StreamReaderProtocol object at 0x7efd46dfcda0>)
DEBUG:asyncio:poll took 0.006 ms: 1 events
True
2
1
# this is the result of can_read()
False
2
1
False
2
8.01068925857544
WARNING:asyncio:Executing <Handle <TaskWakeupMethWrapper object at 0x7efd46e020d8>(<Future finis...events.py:275>) created at /usr/lib/python3.6/asyncio/streams.py:393> took 8.010 seconds
```
By the way, the warning will always print whether i run `can_read` or not(i tried to run the script by make the method call commented and it is still printed)

**Comment by NoneGG** [comment]
@whiteplanet 
You are right, the check of `can_read` is redundant and it will always be false even message is published, i have remove it now. 
I really appreciate your suggestion, that helps a lot~
Please try again with your app now, it should run well.

**Comment by whiteplanet** [comment]
Hi Jason,

Right now when I stop (ctrl+c) the other part of my app (which is actually pushes data into a channel and also uses aredis library), I get another exception: `ResourceWarning: unclosed resource <TCPTransport closed=False reading=True 0x1c504f8>` Looks like some kind of a resource is not released properly.

Also, some times this exception arises in my message listener (when get_message() called):
```
concurrent.futures._base.CancelledError
During handling of the above exception, another exception occurred:

// stack trace …

aredis.exceptions.ConnectionError: Error while reading from stream: ()
During handling of the above exception, another exception occurred:

// stack trace …

aredis/connection.py", line 309, in read_response
    self._reader.feed(buffer)
AttributeError: 'NoneType' object has no attribute 'feed'
```

I'll try to make some little demo tomorrow.

Best,
Jack

**Comment by NoneGG** [comment]
@whiteplanet 

The error in your listener is caused by connection disconnected being used to `get_message` again, this may need check on `self._reader` or related resource (that is what `can_read` check used for before)

About the `TCPTransport`, i agree with your opinion, there seems to be resource not released properly.

Could you give me some demo so i can reproduce and fix the problem? Or maybe i can make it myself tonight.

**Comment by NoneGG** [comment]
@whiteplanet 
I make a publisher and a listener and run them in different process, but neither `TCPTransport` nor  parser of a connection being used after disconnection could be seen in log. I think may it be different scene where your code working. 
My code is like below, could you give me some example to reproduce the problem? 
By the way, could we create a new issue to trace this problem? This may help other people and then they can search the issue.
```Python
# listener
def my_handler(x):
    print(x)


async def listen():
    client = aredis.StrictRedis()
    pubsub = client.pubsub()
    await pubsub.subscribe(**{'my-channel': my_handler})
    while True:
        await pubsub.get_message(timeout=1)
        await asyncio.sleep(0.1)

# publisher
async def publish():
    client = aredis.StrictRedis()
    while True:
        await client.publish('my-channel', 'lalala')
        await asyncio.sleep(0.1)
```

**Comment by whiteplanet** [comment]
Hi Jason,

I've created a test [repo](https://github.com/whiteplanet/aredis-test) and in this test code everything is working fine: no leaks, correct app termination.

I'll double check my project code tomorrow and let you know about results.
Thank you, for the patience and help.

Best,
Jack

**Comment by NoneGG** [comment]
@whiteplanet 
Thank you for help to find the problems. I really appreciate your effort. 
Looking forward to your results~

**Comment by whiteplanet** [comment]
Hi Jason,

I've checked my code and noticed that in test code absent **pubsub** destruction code `pubsub.close()` & unsubscription `pubsub.unsubscribe()` (like in my project code). I've added it to the test project and now the issue appears once again.

Please checkout my test [project](https://github.com/whiteplanet/aredis-test) and run just one listener (listener.py). It now listen for the data only for 3 seconds, then unsubscribe & close the channel, and then exception arises:
```
2017-09-18 16:22:21 - (network)[INFO][192.168.1.217:63243]: POST http://192.168.1.113:8000/rv/v1/auth  200 73
Task exception was never retrieved
future: <Task finished coro=<PubSub._execute() done, defined at /home/jack/.pyenv/rvprint/lib/python3.6/site-packages/aredis/pubsub.py:108> exception=AttributeError("'NoneType' object has no attribute 'feed'",) created at /usr/lib/python3.6/asyncio/tasks.py:345>
source_traceback: Object created at (most recent call last):
  File "startapp.py", line 11, in <module>
    APP.run(host="0.0.0.0", port=8000, debug=APP.config.DEBUG)
  File "/home/jack/.pyenv/rvprint/lib/python3.6/site-packages/sanic/app.py", line 595, in run
    serve(**server_settings)
  File "/home/jack/.pyenv/rvprint/lib/python3.6/site-packages/sanic/server.py", line 527, in serve
    loop.run_forever()
  File "/usr/lib/python3.6/asyncio/coroutines.py", line 125, in send
    return self.gen.send(value)
  File "/home/jack/.pyenv/rvprint/lib/python3.6/site-packages/sanic/app.py", line 503, in handle_request
    response = await response
  File "/usr/lib/python3.6/asyncio/coroutines.py", line 109, in __next__
    return self.gen.send(None)
  File "/home/jack/rvprint/apigate/endpoints.py", line 25, in auth
    context=request)
  File "/usr/lib/python3.6/asyncio/coroutines.py", line 109, in __next__
    return self.gen.send(None)
  File "/home/jack/.pyenv/rvprint/lib/python3.6/site-packages/jsonrpcserver/async_methods.py", line 24, in dispatch
    return await dispatch(self, request, context=context)
  File "/usr/lib/python3.6/asyncio/coroutines.py", line 109, in __next__
    return self.gen.send(None)
  File "/home/jack/.pyenv/rvprint/lib/python3.6/site-packages/jsonrpcserver/async_dispatcher.py", line 49, in dispatch
    return await AsyncRequests(requests).dispatch(methods, context=context)
  File "/usr/lib/python3.6/asyncio/coroutines.py", line 109, in __next__
    return self.gen.send(None)
  File "/home/jack/.pyenv/rvprint/lib/python3.6/site-packages/jsonrpcserver/async_dispatcher.py", line 39, in dispatch
    self.response = await request.call(methods)
  File "/usr/lib/python3.6/asyncio/coroutines.py", line 109, in __next__
    return self.gen.send(None)
  File "/home/jack/.pyenv/rvprint/lib/python3.6/site-packages/jsonrpcserver/async_request.py", line 21, in call
    result = await callable_(*(self.args or []), **(self.kwargs or {}))
  File "/usr/lib/python3.6/asyncio/coroutines.py", line 109, in __next__
    return self.gen.send(None)
  File "/home/jack/rvprint/apigate/authmethods.py", line 22, in _get_params_and_call
    return await func(*args, **kwargs, request=request, app=app)
  File "/usr/lib/python3.6/asyncio/coroutines.py", line 109, in __next__
    return self.gen.send(None)
  File "/home/jack/rvprint/apigate/authmethods.py", line 59, in login
    msg = await rcon.read_message(pbsb, unsubscribe=True)
  File "/usr/lib/python3.6/asyncio/coroutines.py", line 109, in __next__
    return self.gen.send(None)
  File "/home/jack/.pyenv/rvprint/lib/python3.6/site-packages/connections/__init__.py", line 127, in read_message
    message = await self.get_message(pbsb, 0.1)
  File "/usr/lib/python3.6/asyncio/coroutines.py", line 109, in __next__
    return self.gen.send(None)
  File "/home/jack/.pyenv/rvprint/lib/python3.6/site-packages/connections/__init__.py", line 115, in get_message
    return await self._get_message(pbsb, timeout)
  File "/usr/lib/python3.6/asyncio/coroutines.py", line 109, in __next__
    return self.gen.send(None)
  File "/home/jack/.pyenv/rvprint/lib/python3.6/site-packages/connections/__init__.py", line 110, in _get_message
    return await pbsb.get_message(timeout=timeout)
  File "/usr/lib/python3.6/asyncio/coroutines.py", line 109, in __next__
    return self.gen.send(None)
  File "/home/jack/.pyenv/rvprint/lib/python3.6/site-packages/aredis/pubsub.py", line 211, in get_message
    response = await self.parse_response(block=False, timeout=timeout)
  File "/usr/lib/python3.6/asyncio/coroutines.py", line 109, in __next__
    return self.gen.send(None)
  File "/home/jack/.pyenv/rvprint/lib/python3.6/site-packages/aredis/pubsub.py", line 133, in parse_response
    return await asyncio.wait_for(coro, timeout)
  File "/usr/lib/python3.6/asyncio/tasks.py", line 345, in wait_for
    fut = ensure_future(fut, loop=loop)
Traceback (most recent call last):
  File "/home/jack/.pyenv/rvprint/lib/python3.6/site-packages/aredis/connection.py", line 302, in read_response
    buffer = await self._stream.read(self._read_size)
  File "/usr/lib/python3.6/asyncio/streams.py", line 628, in read
    yield from self._wait_for_data('read')
  File "/usr/lib/python3.6/asyncio/streams.py", line 458, in _wait_for_data
    yield from self._waiter
concurrent.futures._base.CancelledError

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/jack/.pyenv/rvprint/lib/python3.6/site-packages/aredis/pubsub.py", line 110, in _execute
    return await command(*args)
  File "/usr/lib/python3.6/asyncio/coroutines.py", line 128, in throw
    return self.gen.throw(type, value, traceback)
  File "/home/jack/.pyenv/rvprint/lib/python3.6/site-packages/aredis/connection.py", line 431, in read_response
    response = await exec_with_timeout(self._parser.read_response(), self._stream_timeout, loop=self.loop)
  File "/usr/lib/python3.6/asyncio/coroutines.py", line 128, in throw
    return self.gen.throw(type, value, traceback)
  File "/home/jack/.pyenv/rvprint/lib/python3.6/site-packages/aredis/connection.py", line 35, in exec_with_timeout
    return await asyncio.wait_for(coroutine, timeout, loop=loop)
  File "/usr/lib/python3.6/asyncio/tasks.py", line 339, in wait_for
    return (yield from fut)
  File "/usr/lib/python3.6/asyncio/coroutines.py", line 128, in throw
    return self.gen.throw(type, value, traceback)
  File "/home/jack/.pyenv/rvprint/lib/python3.6/site-packages/aredis/connection.py", line 306, in read_response
    (e.args,))
aredis.exceptions.ConnectionError: Error while reading from stream: ()

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/usr/lib/python3.6/asyncio/coroutines.py", line 125, in send
    return self.gen.send(value)
  File "/home/jack/.pyenv/rvprint/lib/python3.6/site-packages/aredis/pubsub.py", line 121, in _execute
    return await command(*args)
  File "/usr/lib/python3.6/asyncio/coroutines.py", line 109, in __next__
    return self.gen.send(None)
  File "/home/jack/.pyenv/rvprint/lib/python3.6/site-packages/aredis/connection.py", line 431, in read_response
    response = await exec_with_timeout(self._parser.read_response(), self._stream_timeout, loop=self.loop)
  File "/usr/lib/python3.6/asyncio/coroutines.py", line 109, in __next__
    return self.gen.send(None)
  File "/home/jack/.pyenv/rvprint/lib/python3.6/site-packages/aredis/connection.py", line 35, in exec_with_timeout
    return await asyncio.wait_for(coroutine, timeout, loop=loop)
  File "/usr/lib/python3.6/asyncio/tasks.py", line 339, in wait_for
    return (yield from fut)
  File "/usr/lib/python3.6/asyncio/coroutines.py", line 109, in __next__
    return self.gen.send(None)
  File "/home/jack/.pyenv/rvprint/lib/python3.6/site-packages/aredis/connection.py", line 309, in read_response
    self._reader.feed(buffer)
AttributeError: 'NoneType' object has no attribute 'feed'
```

Best,
Jack

**Comment by NoneGG** [comment]
@whiteplanet 
Thank you for the help! I find the error only take place randomly when using hiredis as parser.

The `CancelledError` is catched by `HiredisParser.read_response` and raised as a `ConnectionError`, which will be retried in error handling by reconnection in `pubsub.parse_response`. Reconnection will invoke `HiredisParser.on_connect`, till now every thing runs well, then connection try to use parser to parse response, code is like below:
```Python
async def read_response(self):
        if not self._stream:
            raise ConnectionError("Socket closed on remote end")

        # _next_response might be cached from a can_read() call
        if self._next_response is not False:
            response = self._next_response
            self._next_response = False
            return response

        response = self._reader.gets()
        while response is False:
            try:
                buffer = await self._stream.read(self._read_size)
            except Exception:
                e = sys.exc_info()[1]
                raise ConnectionError("Error {} while reading from stream: {}".format(type(e), e.args,))
            if not buffer:
                raise ConnectionError("Socket closed on remote end")
            self._reader.feed(buffer)
            response = self._reader.gets()
        if isinstance(response, ResponseError):
            response = self.parse_error(response.args[0])
        return response
```
The problem is, `self._reader` is not None before loop(if it is then `self._reader.gets()` will raise Exception) but after awaken by event loop from `await`, `self._reader` is set to None, and `self._reader.feed(buffer)` raise error. Since connection and parser are in one to one relationship and there is only one connection in use by a pubsub instance, it seems threadsafe. 
I think it may be caused by the unsubscribe after breaking the loop(caused by CancelledError), there may be event not awaken and handled when first CancelledError raised. And when it coming to be handled, connection is disconnected and parser._reader is set to None.(That's why nothing happened without unsubscribed)

All above is my speculation, i am not so sure. But to prove my guess i add `asyncio.sleep` before unsubscribe, no error raised then. What's your opinion?

**Comment by whiteplanet** [comment]
Hello Jason,

This is seem to be working. But could we simply catch `CancelledError` in the `SocketBuffer` class?

```
    async def read_response(self):
        if not self._stream:
            raise ConnectionError("Socket closed on remote end")

        # _next_response might be cached from a can_read() call
        if self._next_response is not False:
            response = self._next_response
            self._next_response = False
            return response

        response = self._reader.gets()
        while response is False:
            try:
                buffer = await self._stream.read(self._read_size)
            except CancelledError:
                break
            except Exception as exp:
                e = sys.exc_info()[1]
                raise ConnectionError("Error while reading from stream: %s" %
                                      (e.args,))
            if not buffer:
                raise ConnectionError("Socket closed on remote end")
            self._reader.feed(buffer)
            response = self._reader.gets()
        if isinstance(response, ResponseError):
            response = self.parse_error(response.args[0])
        return response
```

Best,
Jack

**Comment by whiteplanet** [comment]
I've created a [patch](https://github.com/NoneGG/aredis/files/1314074/cancellederror-patch.txt), please give it a try.

**Comment by NoneGG** [comment]
@whiteplanet 

Thank you for your patch~

I think we should use `raise` instead of `break` in case normal cancellation of coroutine controlled by users. 

Both methods can avoid command being retried twice(if error raised is not instance of `ConnectionError` or `TimeoutError`, error won't be caught in pubsub._execute and ). But if loop `break` directly, `return response` will be executed and `False` will be returned, which will lead result parsing to error of some commands. 

Since the error raise changed, error handling of `StrictRedis` and other clients class are also influenced, there should be a check on the error raised to determine if coroutine should be retried again. 

Would you like to make a pull request and add some comment to explain(it may be better to give the link of this issue) about it in code please?  I will be glad to merge it.

**Comment by whiteplanet** [comment]
Done.
